contracts/svm/programs/express_relay/src/lib.rs

```diff
@@ -309,3 +351,97 @@ pub struct WithdrawFees<'info> {
     #[account(mut, seeds = [SEED_METADATA], bump, has_one = admin)]
     pub express_relay_metadata: Account<'info, ExpressRelayMetadata>,
 }
+
+#[derive(AnchorSerialize, AnchorDeserialize, Eq, PartialEq, Clone, Copy, Debug)]
+pub enum FeeToken {
+    Input,
+    Output,
+}
+
+/// For all swap instructions and contexts, input and output are defined with respect to the searcher
+/// So `mint_input` refers to the token that the searcher provides to the trader and
+/// `mint_output` refers to the token that the searcher receives from the trader
+/// This choice is made to minimize confusion for the searchers, who are more likely to parse the program
+#[derive(AnchorSerialize, AnchorDeserialize, Eq, PartialEq, Clone, Copy, Debug)]
+pub struct SwapArgs {
+    pub amount_input:     u64,
+    pub amount_output:    u64,
+    // The referral fee is specified in basis points
+    pub referral_fee_bps: u64,
+    // Token in which the fees will be paid
+    pub fee_token:        FeeToken,
+}
+
+#[derive(Accounts)]
+#[instruction(data: Box<SwapArgs>)]
+pub struct Swap<'info> {
+    /// Searcher is the party that sends the input token and receives the output token
+    pub searcher: Signer<'info>,
+
+    /// Trader is the party that sends the output token and receives the input token
+    pub trader: Signer<'info>,
+
+    // Searcher accounts
+    #[account(
+        mut,
+        token::mint = mint_input,
+        token::authority = searcher,
+        token::token_program = token_program_input
+    )]
+    pub searcher_input_ta: InterfaceAccount<'info, TokenAccount>,
+
+    #[account(
+        mut,
+        token::mint = mint_output,
+        token::authority = searcher,
+        token::token_program = token_program_output
+    )]
+    pub searcher_output_ta: InterfaceAccount<'info, TokenAccount>,
+
+    // Trader accounts
+    #[account(
+        mut,
+        associated_token::mint = mint_input,
+        associated_token::authority = trader,
+        associated_token::token_program = token_program_input
+    )]
+    pub trader_input_ata: InterfaceAccount<'info, TokenAccount>,
+
+    #[account(
+        mut,
+        associated_token::mint = mint_output,
+        associated_token::authority = trader,
+        associated_token::token_program = token_program_output
+    )]
+    pub trader_output_ata: InterfaceAccount<'info, TokenAccount>,
+
+    // Fee receivers, we can't do any mint checks here because the mint depends on the fee token
+    /// Router fee receiver token account: the referrer can provide an arbitrary receiver for the router fee
```

Comment:
i think if you make this an ata, you can just run all the ATA checks in the context by specifying a fee mint account. otherwise, maybe there's some way to catch the mint based on conditional logic using the instruction args that Anchor exposes?

i'm fine with the way it is rn, doing the ata validation in the code, but i know you wanted to look for ways to do it in the context

Reply:
I moved the validation to the context, it only adds two bytes to the transaction size since it's just 2 extra account indexes

Reply:
I think `router_fee_receiver_ta` as an arbitrary token account works because it gives more flexibility to the receiver

Reply:
`router_fee_receiver_ta` is specified by the referrer when they call the quote api and it should be checked by the server